
# Digit Recognition System
Digit recognition system is the working of a machine to train itself or recognizing the digits from different sources like emails, bank cheque, papers, images, etc. and in different real-world scenarios for online handwriting recognition on computer tablets or system, recognize number plates of vehicles, processing bank cheque amounts, numeric entries in forms filled up by hand (say — tax forms) and so on

# Problems with handwritten digits
The handwritten digits are not always of the same size, width, orientation and justified to margins as they differ from writing of person to person, so the general problem would be while classifying the digits due to the similarity between digits such as 1 and 7, 5 and 6, 3 and 8, 2 and 5, 2 and 7, etc. This problem is faced more when many people write a single digit with a variety of different handwritings. Lastly, the uniqueness and variety in the handwriting of different individuals also influence the formation and appearance of the digits. Now we introduce the concepts and algorithms of deep learning and machine learning.

 The provided Handwritten digits are images in the form of 28*28 gray scale intensities of images representing an image along with the first column to be a label (0 to 9) for every image.

# Loading Neccessary Libraries

In [1]:

import matplotlib.pyplot as plot
import seaborn as sea
from tensorflow.keras.datasets import mnist
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import accuracy_score
from sklearn.tree import DecisionTreeClassifier
from sklearn import preprocessing
import numpy as np
from sklearn.svm import SVC
from google.colab.patches import cv2_imshow
import os,cv2,json,random
from tensorflow import keras
from tensorflow.keras import layers
from sklearn.metrics import precision_score,recall_score,f1_score,roc_auc_score

# Loading Data

In [2]:

(X_train, y_train), (X_test, y_test) = mnist.load_data()


11490434/11490434 [==============================] - 0s 0us/step


In [ ]:
X_train.shape

(60000, 28, 28)

In [ ]:
y_test

array([7, 2, 1, ..., 4, 5, 6], dtype=uint8)

# Using KNN

In [ ]:
# Flatten the images from 28x28 to 784-dimensional vectors
x_train = X_train.reshape(60000, 784)
x_test = X_test.reshape(10000, 784)

In [ ]:
# Scale the pixel values to the range 0-1
x_train = x_train.astype('float32') / 255
x_test = x_test.astype('float32') / 255

In [ ]:
# Create a KNN classifier with a suitable number of neighbors
knn = KNeighborsClassifier(n_neighbors=5)

# Train the classifier on the training data
knn.fit(x_train, y_train)


KNeighborsClassifier()

In [ ]:

# Make predictions on the testing data
y_pred = knn.predict(x_test)

In [ ]:

# Evaluate the accuracy of the model
accuracy = accuracy_score(y_test, y_pred)

print("Accuracy:", accuracy)


Accuracy: 0.9688


## By using KNN model we got accuracy of 96.8%

# Using Decissin Tree

In [ ]:
clf = DecisionTreeClassifier(criterion='gini',max_depth=15,min_samples_leaf=1,min_samples_split=3,splitter='best')
  # Set a random state for reproducibility

# Train the classifier on the training data
clf.fit(x_train, y_train)

# Make predictions on the testing data
y_pred1 = clf.predict(x_test)


In [ ]:
accuracy = accuracy_score(y_test, y_pred)


print("Accuracy:", accuracy)


Accuracy: 0.9688


## By using Decision Tree model we got accuracy of 96.8%

# Using SVM

In [ ]:

# Create an SVM classifier with a suitable kernel
svm = SVC(kernel='rbf', gamma='auto')  # RBF kernel often works well for image data

# Train the classifier on the training data
svm.fit(x_train, y_train)

# Make predictions on the testing data
y_pred = svm.predict(x_test)

# Evaluate the accuracy of the model
accuracy = accuracy_score(y_test, y_pred)
print("Accuracy:", accuracy)

Accuracy: 0.9446


## By using SVM model we got accuracy of 94.46%

# Using CNN

In [18]:
#Loading the Data
img_rows, img_cols = 28, 28

X_train = X_train.reshape(60000, 28, 28, 1)
X_test = X_test.reshape(10000, 28, 28, 1)
# input_shape = (img_rows, img_cols, 1)

# print('input_shape: ', input_shape)
print('x_train shape:', X_train.shape)
print('x_test shape:', X_test.shape)
y_train.shape

x_train shape: (60000, 28, 28, 1)
x_test shape: (10000, 28, 28, 1)


(60000, 10)

In [19]:
# rescale to have values within 0 - 1 range [0,255] --> [0,1]
X_train = X_train.astype('float32')/255
X_test = X_test.astype('float32')/255

print('X_train shape:', X_train.shape)
print(X_train.shape[0], 'train samples')
print(X_test.shape[0], 'test samples')

X_train shape: (60000, 28, 28, 1)
60000 train samples
10000 test samples


In [20]:
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Conv2D, MaxPooling2D, Flatten, Dense, Dropout

# build the model object
model = Sequential()

# CONV_1: add CONV layer with RELU activation and depth = 32 kernels
model.add(Conv2D(32, kernel_size=(3, 3), padding='same',activation='relu',input_shape=(28,28,1)))# 'same' in the sense it'll give the shape same as input
# POOL_1: downsample the image to choose the best features
model.add(MaxPooling2D(pool_size=(2, 2)))

# CONV_2: here we increase the depth to 64
model.add(Conv2D(64,(3, 3),padding='same', activation='relu'))
# POOL_2: more downsampling
model.add(MaxPooling2D(pool_size=(2, 2)))

# flatten since too many dimensions, we only want a classification output
model.add(Flatten())

# FC_1: fully connected to get all relevant data
model.add(Dense(64, activation='relu'))

# FC_2: output a softmax to squash the matrix into output probabilities for the 10 classes
model.add(Dense(10, activation='softmax'))

model.summary()

Model: "sequential_1"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 conv2d_2 (Conv2D)           (None, 28, 28, 32)        320       
                                                                 
 max_pooling2d_2 (MaxPoolin  (None, 14, 14, 32)        0         
 g2D)                                                            
                                                                 
 conv2d_3 (Conv2D)           (None, 14, 14, 64)        18496     
                                                                 
 max_pooling2d_3 (MaxPoolin  (None, 7, 7, 64)          0         
 g2D)                                                            
                                                                 
 flatten_1 (Flatten)         (None, 3136)              0         
                                                                 
 dense_2 (Dense)             (None, 64)               

In [22]:
# compile the model
model.compile(loss='categorical_crossentropy', optimizer='rmsprop',
              metrics=['accuracy'])

In [15]:
from tensorflow.keras.utils import to_categorical
num_classes = 10
# print first ten (integer-valued) training labels
print('Integer-valued labels:')
# print(y_train[:10])

# one-hot encode the labels
# convert class vectors to binary class matrices
y_train = to_categorical(y_train, num_classes)
y_test = to_categorical(y_test, num_classes)


Integer-valued labels:


In [23]:
from tensorflow.keras.callbacks import ModelCheckpoint

# train the model
checkpointer = ModelCheckpoint(filepath='model.weights.best.hdf5', verbose=1,
                               save_best_only=True)
hist = model.fit(X_train, y_train, batch_size=64, epochs=1,
                 validation_data=(X_test, y_test), callbacks=[checkpointer],verbose=2, shuffle=True)





Epoch 1: val_loss improved from inf to 0.04724, saving model to model.weights.best.hdf5
938/938 - 70s - loss: 0.1539 - accuracy: 0.9523 - val_loss: 0.0472 - val_accuracy: 0.9839 - 70s/epoch - 74ms/step


/usr/local/lib/python3.10/dist-packages/keras/src/engine/training.py:3103: UserWarning: You are saving your model as an HDF5 file via `model.save()`. This file format is considered legacy. We recommend using instead the native Keras format, e.g. `model.save('my_model.keras')`.
  saving_api.save_model(


In [24]:
score = model.evaluate(X_test, y_test, verbose=0)
score


[0.04723735153675079, 0.9839000105857849]

## By using CNN  we got accuracy of 98.39%

# Summary:



The Summary here we give is as follows;

- We trained the model using KNN keeping the neighbourhood points as 5. We got accuracy of 96.8%.
- We Trained the model using decission Tree keeping crterion as gini,max depth of 15,minimum sample leaves as 1,minimum samples as 3 and splitter as best.
We got accuracy of 96.8%.
- We trained the model using SVM keeping kernel as rbf since we are using image arrays, it'll be working well and gamma as auto. We got accuarcy of 94.46%.


- The network begins with a sequence of two convolutional layers, followed by max pooling layers.
- The final layer has one entry for each object class in the dataset, and has a softmax activation function, so that it returns probabilities.
- The Conv2D depth increases from the input layer of 1 to 32 to 64.
- We also want to decrease the height and width - This is where maxpooling comes in. Notice that the image dimensions decrease from 28 to 14 after the pooling layer.
- You can see that every output shape has **None** in place of the batch-size. This is so as to facilitate changing of batch size at runtime.
- Finally, we add one or more fully connected layers to determine what object is contained in the image.

 # Conclusion
 - We conclude that the best machine learning  model for the given dataset is KNN and Decission Tree of accuracy 96.8%.
 - But by using Deep learning model we get accuracy of 98.39%.

# Risks that we faced:

- Overfitting- Overfitting occurs when the model learns the training data too well and performs poorly on unseen data. Underfitting occurs when the model is too simple and cannot capture the underlying patterns in the data.
- To overcome these challenges, we have used techniques like regularization, early data splitting to improve the model's performance.